Automattic/sensei
# Add custom slug in module block

Started as a solution to https://github.com/Automattic/sensei/issues/5176 . Also allows adding custom slugs to modules from the module block in the course edit page.

### Changes proposed in this Pull Request

1. On the course edit page, the block settings for the module block will contain a Custom slug section to add a custom slug for the module.
2. If a custom slug is added for a module, it'll be shown below the title of the module so that it's easy to notice and understand that there's a custom slug for that module
3. If the course teacher is changed, the module with the custom slug will belong to the new teacher.

The following 2 are a bit technical:
4. As custom slugs may not have the teacher id prefixed with the slug, we've started keeping the teacher's id as metadata to the module term so that if we need the id of the teacher of the module later (like for showing the name of the module teacher), we can use that meta value.
5. If a module's slug is changed to a custom slug from a conventional slug from the modules page instead of the course block editor page, a hook action saves the teacher id from the slug to the module metadata before changing the slug and losing the data.

### Testing instructions

- In SenseiLMS->Courses, click on 'New Course'. Add a few module blocks.
- Select a module block, and go to the Block settings panel on the right. Add a custom slug.
- That slug will appear below the module title as well. Add custom slug for some, keep some unchanged
- Try saving the course.
- The slugs should persist.
- Go to SenseiLMS->Modules page, change the slug of some modules with custom slugs and some with conventional slug.
- Open the same course again, make sure the edited slugs are visible. Try changing some texts on the editor page and save again. make sure everything is working.
- Change the teacher for the course and save. Make sure the changed teacher appears for all modules correctly.
- Go to modules page again and change the slug of a module that belongs to a teacher. Change the slug so that it does not append the teacher's id anymore.
- Go to the modules page again, make sure that module is still showing the teacher's name.

### Screenshot / Video

https://user-images.githubusercontent.com/6820724/171064046-5ca8eff5-26b5-4b7a-b217-9e06a7750f83.mov
